src/overlays/actors/ovl_Bg_Haka_Ship/z_bg_haka_ship.c

```diff
@@ -41,25 +39,203 @@ const ActorInit Bg_Haka_Ship_InitVars = {
+        this->dyna.actor.world.pos.z = parent->dyna.actor.world.pos.z;
+    } else {
+        this->dyna.actor.parent = NULL;
+    }
+}
+
+void BgHakaShip_WaitForSong(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (Flags_GetSwitch(globalCtx, this->switchFlag)) {
+        if (this->counter) {
+            this->counter--;
+        }
+        if (this->counter == 0) {
+            this->counter = 130;
+            this->actionFunc = BgHakaShip_CutsceneStationary;
+            osSyncPrintf("シーン 外輪船 ...  アァクション！！\n");
+            func_800800F8(globalCtx, 3390, 999, &this->dyna.actor, 0);
+        }
+    }
+}
+
+void BgHakaShip_CutsceneStationary(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (this->counter) {
+        this->counter--;
+    }
+    this->yOffset = sinf(this->counter * (M_PI / 25)) * 6144.0f;
+    if (this->counter == 0) {
+        this->counter = 50;
+        this->actionFunc = BgHakaShip_Moving;
+    }
+}
+
+void BgHakaShip_Moving(BgHakaShip* this, GlobalContext* globalCtx) {
+    f32 distanceFromHome;
+    BgHakaShip* child;
+
+    if (this->counter) {
+        this->counter--;
+    }
+    if (this->counter == 0) {
+        this->counter = 50;
+    }
+    this->dyna.actor.world.pos.y = (sinf(this->counter * (M_PI / 25)) * 50.0f) + this->dyna.actor.home.pos.y;
+
+    distanceFromHome = this->dyna.actor.home.pos.x - this->dyna.actor.world.pos.x;
+    if (distanceFromHome > 7650.0f) {
+        this->dyna.actor.world.pos.x = this->dyna.actor.home.pos.x - 7650.0f;
+        this->dyna.actor.speedXZ = 0.0f;
+    }
+    if ((distanceFromHome > 7600.0f) && !Gameplay_InCsMode(globalCtx)) {
+        this->counter = 40;
+        this->dyna.actor.speedXZ = 0.0f;
+        func_8010B680(globalCtx, 0x5071, NULL);
+        this->actionFunc = BgHakaShip_SetupCrash;
+    } else {
+        Math_StepToF(&this->dyna.actor.speedXZ, 4.0f, 0.2f);
+    }
+    child = (BgHakaShip*)this->dyna.actor.child;
+    if (child && child->dyna.actor.update) {
+        child->dyna.actor.shape.rot.z += ((655.0f / 13.0f) * this->dyna.actor.speedXZ);
+    } else {
+        this->dyna.actor.child = NULL;
+    }
+    this->yOffset = sinf(this->counter * (M_PI / 25)) * 6144.0f;
+}
+
+void BgHakaShip_SetupCrash(BgHakaShip* this, GlobalContext* globalCtx) {
+    if (this->counter) {
+        this->counter--;
+    }
+    if (this->counter == 0) {
+        this->counter = 40;
+        this->actionFunc = BgHakaShip_CrashShake;
+    }
+    Math_ScaledStepToS(&this->yOffset, 0, 128);
+}
+
+void BgHakaShip_CrashShake(BgHakaShip* this, GlobalContext* globalCtx) {
+    DECR(this->counter);
+    this->dyna.actor.world.pos.y = this->counter % 4 * 3 - 6 + this->dyna.actor.home.pos.y;
+    if (!this->counter) {
+        this->dyna.actor.gravity = -1.0f;
+        this->actionFunc = BgHakaShip_CrashFall;
+    }
+    func_8002F974(&this->dyna.actor, 0x205C);
+}
+
+void BgHakaShip_CrashFall(BgHakaShip* this, GlobalContext* globalCtx) {
+    BgHakaShip* child;
```

Comment:
Like before this could probably be `Actor*` but also like before wait for others to input.